vertical-pod-autoscaler/enhancements/4831-control-actuation-behavior/README.md

```diff
@@ -0,0 +1,105 @@
+of type `[]*ActuationRequirement`. A single `ActuationRequirement` defines a condition which must be `true` to allow
+actuation for the corresponding `Container`. When multiple `ActuationRequirements` are specified for a `Pod`, all of
+them must evaluate to `true` to allow eviction. VPA may partially actuate its recommendation in place, obeying all the
+`ActuationRequirements` while not applying parts of the recommendation that would violate any `ActuationRequirement`.
+
+A single `ActuationRequirement` specifies `Resources` and a `ChangeRequirement` comparing the new recommendation
+(`Target`) with the existing requests on a Pod (`Requests`). Possible values for `Resources` are `[CPU]` and `[Memory]`
+or both `[CPU,Memory]`. If `Resources: [CPU, Memory]`, the condition must be true for either of the two resources to
+allow for actuation. Possible values for `ChangeRequirement` are `TargetHigherThanRequests`,
+`TargetHigherThanOrEqualToRequests`, `TargetLowerThanRequests` and `TargetLowerThanOrEqualToRequests`.
+
+Add validation to prevent users from adding `ActuationRequirements` which can never evaluate to `true`:
+* Reject if more than one `ActuationRequirement` for a single resource is found
+* Reject if `Resource: [CPU, Memory]` is specified on one `ActuationRequirement` together with `Resource: [CPU]` or
+* `Resource: [Memory]` on another `ActuationRequirement`
+
+## Design Details
+### Test Plan
+* Add automated E2E tests verifying that VPA only evicts when all `ActuationRequirements` evaluate to `true`
+* Keep existing tests which verify that nothing changes when users don't specify any `ActuationRequirements`
+* Add tests verifying the validation
+
+## Alternatives
+### Run VPA in recommendation only mode
+We currently achieve a similar functionality by running VPA in `spec.updatePolicy.updateMode: "Off"` and having a 
+different controller inspect the recommendations and apply them based on certain criteria like scaling direction and 
+time.
+With this approach you either end up re-building half of the VPA (updater/webhook), or use a different mechanism to
+apply the recommendations, such as modifying the requests in the Pod owning Object – which has its own drawbacks.
+
+### Independent controls for eviction and in-place actuation
+We could introduce independent fields controlling scaling by eviction and scaling in place. This seems redundant.
+
+The following analysis focuses on the `InPlaceOrRecreate` mode. In other modes having separate controls for in-place and
+eviction actuation doesn't offer any more control than single setting, because VPA won't actively actuate (`Off` and
+`Initial`) or because it will use only one mode of actuation (`Recreate` and `InPlaceOnly`). It also assumes that scale
+downs always succeed.
+
+VPA wil attempt to scale in place before attempting to evict. Since scale downs succeed VPA will never consider evicting
```

Comment:
Done

> > So, when in-place resource updates make it to the VPA, I don't think I would want any restrictions being applied to disruption-free resource updates and _still_ want to have restrictions for eviction-based resource updates – be it because VPA is falling back to eviction or because my workload is configured with an update mode which doesn't allow for in-place updates.
> 
> Marco Voelz , I'm not sure I fully understood here: are you saying you'd always want in-place updates to happen and only care about evictions?
> 
> If yes: unlimited in-place scale down is likely to work but might end up in an eviction later as it might not be reversible (e.g. more Pods getting scheduled on a Node or others growing). Would that be a good reason do limit in-place scale down as well?

@voelzmo @kgolab 

I don't think eviction will go away completely in the foreseeable future. I don't think we'll have guarantee that in-place update will succeed so we'll want eviction as a fallback.

This is more about the case Karol described - user might want to avoid scale downs (because there is no guarantee that they'll be able to scale up without disruption) but allow scale-ups (preferring in-place but with eviction as fallback).